### PR TITLE
Make Settings usable in .NET scenarios

### LINKED ISSUE #6784: Updating the Default Settings file in a WinForms .NET 5 Visual Basic app fails.
**Visual Studio Version**:
\>=16.8
**Summary**:
Attempting to save after changing a value in the Default Settings file of a WinForms .NET 5 Visual Basic app fails.

**Steps to Reproduce**:

1. Create a new .NET 5 WinForms App.
2. Insert a new App.Config to the project by chosing Add/New Items and General/Application Configuration File.
3. Show All Files in the Solution Explorer and select the My Project branch.
4. Insert a new Settings file to the project by choosing Add/New Items and General/Settings File. Make sure to rename the Name to _Settings.settings_.
5. In the Settings editor, which appears now, add a value and try to save the settings.

**Expected Behavior**:
Settings file should be written, settings designer code be generated.
**Actual Behavior**:
image

Exception:
'An error occurred creating the configuration section handler for userSettings/VbWinForms5EventTest.Settings: Could not load file or assembly 'System.Configuration.ConfigurationManager, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.'

**User Impact**:
Interactively using the settings designer in a .NET 5 VB WinForms App is not possible.

Stack Trace:
```
   at System.Configuration.BaseConfigurationRecord.FindAndEnsureFactoryRecord(String configKey, Boolean& isRootDeclaredHere)
   at System.Configuration.BaseConfigurationRecord.GetSectionRecursive(String configKey, Boolean getLkg, Boolean checkPermission, Boolean getRuntimeObject, Boolean requestIsHere, Object& result, Object& resultRuntimeObject)
   at Microsoft.VisualStudio.Shell.Design.Serialization.ConfigurationHelperService.GetConfigSection(Configuration config, String sectionName, Boolean isUserScoped, Boolean declare)
   at Microsoft.VisualStudio.Shell.Design.Serialization.ConfigurationHelperService.ReadSettings(ExeConfigurationFileMap fileMap, ConfigurationUserLevel userLevel, DocData appConfigDocData, String sectionName, Boolean isUserScoped, SettingsPropertyCollection settings)
   at Microsoft.VisualStudio.Editors.SettingsDesigner.AppConfigSerializer.SynchronizeUserConfig(String SectionName, IVsHierarchy Hierarchy, ConfigurationHelperService ConfigHelperService, Dictionary`2 SettingsTheDesignerKnownsAbout, DocData AppConfigDocData) in D:\Git\project-system\src\Microsoft.VisualStudio.Editors\SettingsDesigner\AppConfigSerializer.vb:line 402
   at Microsoft.VisualStudio.Editors.SettingsDesigner.AppConfigSerializer.SynchronizeUserConfig(String SectionName, IVsHierarchy Hierarchy, ConfigurationHelperService ConfigHelperService, DesignTimeSettings Settings, DocData AppConfigDocData) in D:\Git\project-system\src\Microsoft.VisualStudio.Editors\SettingsDesigner\AppConfigSerializer.vb:line 347
   at Microsoft.VisualStudio.Editors.SettingsDesigner.AppConfigSerializer.Serialize(DesignTimeSettings Settings, SettingsTypeCache typeCache, SettingsValueCache valueCache, String SectionName, DocData AppConfigDocData, IVsHierarchy Hierarchy, Boolean ShouldSynchronizeUserConfig) in D:\Git\project-system\src\Microsoft.VisualStudio.Editors\SettingsDesigner\AppConfigSerializer.vb:line 320
```

### PR DESCRIPTION
**Pull request by RussKie**
### Changes

1. Pin configuration type names to the desktop identities so not only we can create necessary configuration sections and serialize data, but also that we can deserialize those as well.
Affected types:
    - `System.Configuration.ApplicationSettingsGroup`
    - `System.Configuration.ClientSettingsSection`
    - `System.Configuration.UserSettingsGroup`

    Fixes #6784
    Resolves dotnet/winforms#4308

2. Remove ability to browse and serialize arbitrary types, and restrict the set of type to the the known list

    Resolves https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1388857/

### Tests

Manual using a multi-project solution including
- .NET Framework C# 4.0 (Console) and C# 4.7.2 (WinForms), 
- .NET C# (WinForms) 3.1, 5.0, 6.0
- .NET VB (WinForms) 5.0, 6.0

https://user-images.githubusercontent.com/4403806/135218957-ae46a104-67d6-4b9c-a6d3-d6948deabdec.mp4

/cc: @dotnet/dotnet-winforms 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7647)

### CODE REVIEW

File: src/Microsoft.VisualStudio.Editors/SettingsDesigner/SettingsDesignerView.vb [hunk 1]
```diff
@@ -413,6 +414,12 @@ Namespace Microsoft.VisualStudio.Editors.SettingsDesigner
                 _projectSystemSupportsUserScope = True
             End If
 
+            ' Do not allow browsing or serializing arbitrary types for .NET scenarios
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.Editors/SettingsDesigner/SettingsDesignerView.vb:417]:
```suggestion
            ' Do not allow browsing or serializing arbitrary types for .NET Core scenarios.
            ' We don't currently have a general mechanism to identify types known to both the
            ' designer (running on .NET Framework) and the application (running on .NET Core).
```

File: src/Microsoft.VisualStudio.Editors/SettingsDesigner/SettingsTypeCache.vb [hunk 2]
```diff
@@ -137,6 +137,16 @@ Namespace Microsoft.VisualStudio.Editors.SettingsDesigner
             Dim qualifiedAssemblyName As String = Nothing
 
             If Not String.IsNullOrEmpty(sourceTypeName) Then
+
+                ' The configuration types represent a stable contract, for .NET scenarios pin all these type names to the desktop identities.
+                ' This will ensure that when we create required sections in app.config we'll be able to read them back.
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.Editors/SettingsDesigner/SettingsTypeCache.vb:142]:
```suggestion
                ' The configuration types represent a stable contract, for .NET scenarios pin all these type names to the .NET Framework identities.
                ' This will ensure that when we create required sections in app.config we'll be able to read them back.
                ' Note that when the designer moves off of .NET Framework we may need to revisit this.
```

**Comment by tmeschter** [comment]
This seems like a very reasonable approach to get the designer basically functional for Dev17.0 as we look into more general solutions for 17.1.

@RussKie Did you test with a project that multi-targets to both .NET Framework and .NET Core (e.g. .NET Framework 4.7.2 and .NET Core 5.0)? If so, did you test with different orderings of items in `<TargetFrameworks>`? E.g., try both

`<TargetFrameworks>net472;net5.0</TargetFrameworks>`

and 

`<TargetFrameworks>net5.0;net472</TargetFrameworks>`

There are, unfortunately, times when the order matters. :-(

Also, did you verify this still works with a non-SDK-style project targeting .NET Framework?

**Comment by KlausLoeffelmann** [comment]
This is really cool! Also @KathleenDollard FYI!

**Comment by merriemcgaw** [comment]
> There are, unfortunately, times when the order matters. :-(

Yeah, that's a frequent problem for designer users. We can only display one designer at a time - so when a .NET Framework TFM is loaded, we are only working with the .NET Framework designer and everything builds etc as it did before. I would definitely want to check that the .NET Framework stuff isn't impacted, though it shouldn't be. We should check that also for CSProj, good call.

**Comment by RussKie** [comment]
> Did you test with a project that multi-targets to both .NET Framework and .NET Core (e.g. .NET Framework 4.7.2 and .NET Core 5.0)?

No, forgot about this. Will check 👍 



> did you verify this still works with a non-SDK-style project targeting .NET Framework?

Yes, .NET Framework C# 4.0 (Console). No observable changes.

**Comment by RussKie** [comment]
I've created more test apps including those targeting .NET Framework 2.0, 3.5, 4.0, 4.7.2, .NET Core 3.1, .NET 5.0 and 6.0. As well as MT scenarios (e.g. `net35,net472,net6.0-windows` and `net6.0-windows;net472`). Everything appears to be working expected.

The only unexpected behaviour I observed was that for `net35,net472,net6.0-windows` scenario settings were written as 4.0.0.0 and not 2.0.0.0, but this is an existing behaviour:
image

There's probably a bug, but I guess this scenario is probably in the realms of unsupported edge cases.


Test solution: [ProjectSystemTestHarness.zip](https://github.com/dotnet/project-system/files/7256289/ProjectSystemTestHarness.zip)

**Review by tmeschter — APPROVED** [review]

**Comment by tmeschter** [comment]
Merging. FYI @RussKie @merriemcgaw @jjmew We'll need to take this through M2 approval and possibly QB approval.

**Comment by RussKie** [comment]
Thank you 🎉

**Comment by merriemcgaw** [comment]
@Pilchie as FYI in case  I'm still oof sick tomorrow.

**Comment by MiYanni** [comment]
@Pilchie This was merged into main on Oct 1st. Main (at that point) was still for 17.0 changes, so these changes should be in 17.0. Insertion here: https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/355436

**Comment by Pilchie** [comment]
Thanks for confirming!
